x/expiration/types/expiration.go

```diff
@@ -0,0 +1,109 @@
+package types
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func NewExpiration(
+	moduleAssetID string,
+	owner string,
+	time time.Time,
+	deposit sdk.Coin,
+	message types.Any,
+) *Expiration {
+	return &Expiration{
+		ModuleAssetId: moduleAssetID,
+		Owner:         owner,
+		Time:          time,
+		Deposit:       deposit,
+		Message:       message,
+	}
+}
+
+// ValidateBasic basic format checking of the data
+func (e *Expiration) ValidateBasic() error {
+	now := time.Now()
+	if strings.TrimSpace(e.ModuleAssetId) == "" {
+		return ErrEmptyModuleAssetID
+	}
+	if err := validateAddress(e.ModuleAssetId); err != nil {
+		return err
+	}
+	if strings.TrimSpace(e.Owner) == "" {
+		return ErrEmptyOwnerAddress
+	}
+	if e.Time.Before(now) {
+		return ErrTimeInPast
+	}
+	if !e.Deposit.IsValid() {
+		return ErrInvalidDeposit
+	}
+	if err := e.validateMessage(); err != nil {
+		return ErrInvalidMessage
+	}
+	return nil
+}
+
+// validateMessage validates `expiration.Message` conforms to `sdk.Msg`
+// and is registered and whitelisted with the InterfaceRegistry
+func (e *Expiration) validateMessage() error {
+	// this may occur in message decoding
+	if e.Message.Value == nil {
+		return fmt.Errorf("expecting non nil value to validate an Any message")
+	}
+	if e.Message.TypeUrl == "" {
+		return fmt.Errorf("expecting message type URL to unpack Any")
+	}
+	// validate message is a whitelisted sdk.Msg during a BroadcastTx
+	var msg sdk.Msg
+	if err := ModuleCdc.UnpackAny(&e.Message, &msg); err != nil {
+		return err
+	}
```

Comment:
Is there even a reason we need to whitelist message types?

Reply:
Yes. I noticed that tests fail where fail after a `tx` broadcast because it could not find the interfaces to do the unmarshalling. I didn't know how else to handle it.

Reply:
I was asking why we should have a whitelist and this piece of validation at all. In other words, why do we need to restrict messages used in an expiration? What's the downside to not having this piece of validation?

The codec in the keeper should have all message types registered. So it should be able to unpack any message just fine. But this validation here is making sure that the message type is in a smaller whitelist of message types, and it's not clear to me why we need that.

The stuff in the `types` package should be mostly static stuff. The `keeper` package is for stuff that can change based on setup. We really really don't want the expiration module to have dependencies on modules that are going to use it. Specifically, in this case, the `x/expiration` module should not have any imports to anything in the `x/metadata` module. And `types` is not the place to store a whitelist that is built as part of app setup. So this piece of validation isn't really suitable for a `ValidatBasic()`.

This piece of validation should be split out to it's own function that takes in a codec. Then, in the keeper (where we've got a codec and maybe a whitelist codec), make a call to that function as part of further validation. Then, if we decide we still want a whitelist, provide the whitelist codec, otherwise, just provide the full codec.

Reply:
marking this as resolved since we talked about a different approach in https://github.com/provenance-io/provenance/pull/1115#discussion_r988299600. 

We're safe to remove the `UnpackAny` check because new expirations are only created from other modules.